keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md

```diff
@@ -0,0 +1,891 @@
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+When a cluster has multiple apiservers at mixed versions (such as during an
+upgrade or downgrade), not every apiserver can serve every resource at every
+version.
+
+To fix this, we will add a filter to the handler chain in the aggregator which
+proxies clients to an apiserver that is capable of handling their request.
+
+## Motivation
+
+When an upgrade or downgrade is performed on a cluster, for some period of time
+the apiservers are at differing versions and are able to serve different sets of
+built-in resources (different groups, versions, and resources are all possible).
+
+In an ideal world, clients would be able to know about the entire set of
+available resources and perform operations on those resources without regard to
+which apiserver they happened to connect to. Currently this is not the case.
+
+Today, these things potentially differ:
+* Resources available somewhere in the cluster
+* Resources known by a client (i.e. read from discovery from some apiserver)
+* Resources that can be actuated by a client
+
+This can have serious consequences, such as namespace deletion being blocked
+incorrectly or objects being garbage collected mistakenly.
+
+### Goals
+
+* Ensure discovery reports the same set of resources everywhere (not just group
+  versions, as it does today)
+* Ensure that every resource in discovery can be accessed successfully
+* In the failure case (e.g. network not routable between apiservers), ensure
+  that unreachable resources are served 503 and not 404.
+
+### Non-Goals
+
+* Lock particular clients to particular versions
+
+## Proposal
+
+We will use the existing `StorageVersion` API to figure out which group, versions,
+and resources an apiserver can serve.
+
+
+API server change:
+* A new handler is added to the stack:
+
+  - If the request is for a group/version/resource the apiserver doesn't have
+    locally (we can use the StorageVersion API), it will proxy the request to
```

Comment:
a kube-apiserver would lookup the known StorageVersions informer/lister or it would use an internal representation that isn't latent?

Reply:
kube-apiserver would lookup the StorageVersions informer to populate an internal map that would store the list of apiservers that can serve a storageversion. Details here: https://github.com/kubernetes/enhancements/blob/b4cdb22bbe5adb358009b20df4f9a81e7f3a38b4/keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md#aggregation-layer